src/lib/ffi/ffi_tpm2.cpp

```diff
@@ -0,0 +1,89 @@
+/*
+* (C) 2024 Jack Lloyd
+* (C) 2024 René Meusel, Amos Treiber - Rohde & Schwarz Cybersecurity GmbH
+*
+* Botan is released under the Simplified BSD License (see license.txt)
+*/
+
+#include <botan/ffi.h>
+
+#include <botan/internal/ffi_rng.h>
+#include <botan/internal/ffi_util.h>
+
+#if defined(BOTAN_HAS_TPM2)
+   #include <botan/tpm2.h>
+   #include <botan/tpm2_rng.h>
+#endif
+
+extern "C" {
+
+using namespace Botan_FFI;
+
+#if defined(BOTAN_HAS_TPM2)
+/**
+ * This wrapper is required since BOTAN_FFI_DECLARE_STRUCT internally produces a unique pointer,
+ * but the TPM2_Context is meant to be used as a shared pointer.
+ */
+struct botan_tpm2_ctx_wrapper {
+      std::shared_ptr<Botan::TPM2_Context> ctx;
+};
+
+BOTAN_FFI_DECLARE_STRUCT(botan_tpm2_ctx_struct, botan_tpm2_ctx_wrapper, 0xD2B95E15);
+#endif
+
+int botan_tpm2_ctx_init(botan_tpm2_ctx_t* ctx_out, const char* tcti_nameconf) {
+#if defined(BOTAN_HAS_TPM2)
+   return ffi_guard_thunk(__func__, [=]() -> int {
+      if(ctx_out == nullptr) {
+         return BOTAN_FFI_ERROR_NULL_POINTER;
+      }
+      auto ctx = std::make_unique<botan_tpm2_ctx_wrapper>();
+
+      auto tcti = [=]() -> std::optional<std::string> {
+         if(tcti_nameconf == nullptr) {
+            return {};
+         } else {
+            return std::string(tcti_nameconf);
+         }
+      }();
+
+      ctx->ctx = Botan::TPM2_Context::create(std::move(tcti));
```

Comment:
Good point. I don't think reusing `BOTAN_FFI_ERROR_NOT_IMPLEMENTED` is all that helpful since the function would be implemented and TPM2 usage desired in that case (as the build is configured with `--with-tpm2`).

The exception that would be thrown here is `TPM2_Error` with error type `ErrorType::TPMError`. I added a corresponding FFI error type such that the caller will now get that error translated within `ffi_guard_thunk` instead of `BOTAN_FFI_ERROR_EXCEPTION_THROWN`. That way, one can distinguish between a runtime TPM error when configured with TPM and a not implemented error when TPM is not configured.